### ISSUE TITLE
functional api support

### ISSUE THREAD

**Issue by Kingwl** [wonderful-panda/vue-tsx-support#35]
https://github.com/vuejs/vue-function-api

**Comment by u3u** [comment]
The [`ofType<Props>().covert`]() method should support [`ComponentOptions`](https://github.com/vuejs/vue/blob/b7c2d9366cf731a1551286b8ac712e6e0905070e/types/options.d.ts#L67-L123) and [`FunctionalComponentOptions`](https://github.com/vuejs/vue/blob/b7c2d9366cf731a1551286b8ac712e6e0905070e/types/options.d.ts#L124-L145) and should even support [pure functions](https://github.com/vuejs/jsx/tree/dev/packages/babel-sugar-functional-vue#details) that return `VNode | VNode[]`.

```tsx
import { RenderContext } from 'vue';
import { ofType } from 'vue-tsx-support';

export interface HelloProps {
  msg: string;
}

const Hello = ({ props }: RenderContext<HelloProps>) => (
  <div class="hello">{props.msg}</div>
);

// Now you need to use `as any` or `// @ts-ignore` to ignore type checking.
export default ofType<HelloProps>().convert(Hello as any);
```

```tsx
import Vue from 'vue';
import { createComponent, value, onMounted } from 'vue-function-api';
import { ofType } from 'vue-tsx-support';

interface Inject {
  count: number;
}

const Counter = createComponent({
  setup() {
    const count = value(0);

    onMounted(() => {
      window.setTimeout(() => {
        count.value++;
      }, 3000);
    });

    return { count };
  },

  render(this: Vue & Inject) {
    const { count } = this;
    return <div class="counter">{count}</div>;
  },
});

// @ts-ignore
export default ofType().covert(Counter);
```

**Comment by wonderful-panda** [comment]
I'm still considering how should we support composition-api.
This is workaround for now.

1. setup [@vue/composition-api](https://github.com/vuejs/composition-api) and [babel-preset-vca-jsx](https://github.com/luwanquan/babel-preset-vca-jsx) according to their READMEs.

2. Put this code into your project. (for explanation, save as `vca.ts`)

```typescript
import Vue, { VNode } from "vue";
import {
  PropsForOutside,
  RequiredPropNames,
  TsxComponent
} from "vue-tsx-support";
import { createComponent } from "@vue/composition-api";
import { RecordPropsDefinition } from "vue/types/options";
import { SetupContext } from "@vue/composition-api";

export interface CompositionComponentOptions<
  Props,
  PropsDef extends RecordPropsDefinition<Props>
> {
  name?: string;
  props?: PropsDef;
  setup(this: void, props: Props, ctx: SetupContext): () => VNode;
}

export interface ComponentFactory<EventsOn> {
  create<
    Props,
    PropsDef extends RecordPropsDefinition<Props>,
    RequiredProps extends keyof Props = RequiredPropNames<PropsDef> &
      keyof Props
  >(
    options: CompositionComponentOptions<
      Props,
      PropsDef & RecordPropsDefinition<Props>
    >
  ): TsxComponent<Vue & Props, PropsForOutside<Props, RequiredProps>, EventsOn>;
}

const factory = {
  create(options: any): any {
    return createComponent(options);
  }
} as ComponentFactory<{}>;

export const component = factory.create;

export function componentFactoryOf<EventsOn>(): ComponentFactory<EventsOn> {
  return factory;
}
```

And you can use composition-api with TSX like below

```typescript
import { ref, onMounted } from "@vue/composition-api";
import * as vca from "./vca";

export const Counter = vca.component({
  props: {
    initialValue: { type: Number, default: 0 }
  },
  setup(props) {
    const count = ref(props.initialValue);
    onMounted(() => {
      window.setTimeout(() => {
        count.value++;
      }, 3000);
    });
    return () => <div>{count.value}</div>;
  }
});
```

**Comment by dvic** [comment]
@wonderful-panda maybe it's better to implement the support directly in `@vue/composition-api`? I can imagine a lot of people would be happy about this :)

**Comment by mesqueeb** [comment]
@wonderful-panda Two things I wanted to ask:
1. Maybe we should rename this issue to `composition-api` 😉 
2. I'm confused about the difference between `vue-tsx-support` and `babel-preset-vca-jsx`, are they both Vue compilers to convert TSX files into regular Vue files? Is there anywhere I can read more information about that?

**Comment by wonderful-panda** [comment]
I've released 3.0.0-beta.7 which works with @vue/composition-api and babel-preset-vca-jsx.

**Comment by mesqueeb** [comment]
@wonderful-panda I'm thinking of upgrading to v3 to be able to use the composition api.

I see it's still under heavy development though. When you think you think is a good time to upgrade? I want to avoid my project from crashing when I upgrade all package minor updates. : D
